src/bot.js

```diff
@@ -69,29 +63,28 @@ class HangoutsChatBot extends Adapter {
       thread = undefined,
       text = '',
       cardString = '[]',
-      httpRes = undefined) {
+      httpRes = undefined,
+      done = ()=>{}) {
     if (text == '' && cardString == '[]') {
       throw new Error('You cannot send an empty message.');
     }
+    const data = this.mapToGoogleChatResponse(space, text, cardString, thread);
+    this.robot.logger.info('Sending a message to space: ' + space);
+    this.createMessageUsingRestApi_(space, data, done);
+  }
 
+  mapToGoogleChatResponse(space, text, cardString, thread) {
     let data = {
       space: {
         name: space,
       },
       text,
       cards: JSON.parse(cardString),
     };
-
     if (thread) {
       data.thread = thread;
     }
-
-    this.robot.logger.info('Sending a message to space: ' + space);
-    if (httpRes) {
-      httpRes.json(data);
-    } else {
-      this.createMessageUsingRestApi_(space, data);
-    }
```

Comment:
Please continue sending the request using `httpRes.json(data);` for replies if it is present. This way, HTTP bots can still use `res.reply()` without needing a service account.

Reply:
If the intent is that `resp.reply()` from a script results in replying back to the initial request, I'd like to design that a little bit differently. Let me take a stab at that.

Reply:
Thinking out loud some more, the response from the script is executed asynchronously so is it really worth having code that writes to the original req/res stream if it's going to be executed asynchronously anyway?

Reply:
Can you please elaborate on what a service account is in this context and why one is needed to use `res.reply()`?

Reply:
So here's the problem I'm running into. Getting `ERROR Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client` when calling `httpRes.json(data)`. I think it's because `reply` is being executed asynchronously via `process.nextTick()`.

Reply:
Another issue with a design that writes to the response, if multiple scripts `reply` to the same message, the design would have to build an array of responses before actually sending the response.

Reply:
I think the scenario where multiple scripts respond to the same message requires a design where all of the responses sent independent of each other, not on the same req/resp stream.

Reply:
> Can you please elaborate on what a service account is in this context and why one is needed to use res.reply()?

A service account is required to use the Hangouts Chat API. Such an account is not needed for the req/res stream, so currently HTTP bots can forego setting up a service account and still be able to use `res.reply()`. 

> if multiple scripts reply to the same message, the design would have to build an array of responses before actually sending the response.

What about using the Chat API for replies beyond the first? The code would look like:

```
if (httpRes && !httpRes.headersSent)
  httpRes.json(data);
else
  this.createMessageUsingRestApi(data);
```

Reply:
Wouldn't that defeat the intent of not requiring a google service account?

Reply:
Another issue is that in the adapters constructor, it automatically tries to sign into google chat with credentials and if there isn't a valid service account, it'll throw errors.

Reply:
Ok. I've pushed a design where multiple scripts can handle the same message and reply to it, resulting in those messages being sent as an Array in the response.

Reply:
> Wouldn't that defeat the intent of not requiring a google service account?

The majority of bot interactions involve a single response to a given command, this is just patching up an edge case.

> if there isn't a valid service account, it'll throw errors.

This only prevents you from using the service account. We should probably change that message to only be an error for Pub/Sub and a warning otherwise.

> resulting in those messages being sent as an Array in the response.

Unfortunately, Hangouts Chat only supports synchronous responses containing a single message, so an array-based approach won't work. I think sending the first reply over HTTP and remaining ones asynchronously through the API is the best approach here.

Reply:
That seems to defeat the purpose then. Necessitating a service account for the asynchronous messages. In that case, just make them all asynchronous. Am I missing something here?

Reply:
Sure, let's do that then. The main reason for using synchronous responses is for backward compatibility, but this is in a new major version and adding a service account is easy, so it isn't that bad.

Reply:
After re reading your responses, I think I understand this better. Please correct me if I'm wrong. The assumption you're making is that people will typically ensure that there will be 1 script handling 1 message. If we make this assumption and create a design that assumes it, I want to at least log a warning if there's multiple scripts that handle a single message just to have a better UX for the developer.

Reply:
Yes, I'm making that assumption. Logging a warning sounds good to me.

Reply:
Ok. I've updated the design for this scenario. The first response is sent to the original request. The subsequent responses are sent via the `send` method.

Reply:
welp, hold on. gotta defend against `res=null`

Reply:
Ok. I've added a check for `isPubSub`.

Reply:
Before this is accepted, how do you feel about having 2 adapters, one for HTTP and another for Pub Sub and `main.js` is responsible for choosing which one to `use`?

Reply:
> The subsequent responses are sent via the send method.

`send` always posts in a new thread. Since `reply` posts in the same thread when possible, we can't use the `send` method for `reply`. Would you be able to still use the model of send/reply letting `postMessage_` do all the work and perform the `headersSent` check there?

> Before this is accepted, how do you feel about having 2 adapters, one for HTTP and another for Pub Sub and main.js is responsible for choosing which one to use?

Personally, I think that the difference between the two flows is too small to warrant having two separate adapters, but I'm open to the idea of splitting them if you have a nice split in mind.

Reply:
I'll make the change to use postMessage_. I'll create a different pull request for separating pub/sub and http.

Reply:
I've pushed the change.

Reply:
Sorry to go back and forth on this, but I spoke with Daniel and I think that always using asynchronous responses (i.e. using the API) is better. That way, we also wouldn't need two adapters. Could you please make that change? 

Also, we might want to do a major version bump to 3.0.0 since it might break existing HTTP bots.

Reply:
No worries on the back and forth. We're collaborating :)

Are you saying that I should change the code to not use `http.json(data)`? and instead have `reply` execute `postMessage`?

Totally agree about it being a major version change. I'll update the version.

Reply:
Yep, stop using http.json(data) and always use the API instead. Thanks!

Reply:
Ok. I'm going to press resolve convo....